tests/binary/binaries_test.go

```diff
@@ -39,46 +39,84 @@ func TestMain(m *testing.M) {
-				// Currently istiod is not using CLI flags - version/etc will be included in a
-				// detached file.
-				return
-			}
+	runBinariesTest(t, func(t *testing.T, name string) {
+		cmd := path.Join(*releasedir, name)
+		args := []string{"version", "-ojson"}
+		if name == "istioctl" {
+			args = append(args, "--remote=false")
+		}
 
-			args := []string{"version", "-ojson"}
-			if b == "istioctl" {
-				args = append(args, "--remote=false")
-			}
+		out, err := exec.Command(cmd, args...).Output()
+		if err != nil {
+			t.Fatalf("--version failed with error: %v. Output: %v", err, string(out))
+		}
 
-			out, err := exec.Command(cmd, args...).Output()
-			if err != nil {
-				t.Fatalf("--version failed with error: %v. Output: %v", err, string(out))
-			}
+		var resp version.Version
+		if err := json.Unmarshal(out, &resp); err != nil {
+			t.Fatalf("Failed to marshal to json: %v. Output: %v", err, string(out))
+		}
 
-			var resp version.Version
-			if err := json.Unmarshal(out, &resp); err != nil {
-				t.Fatalf("Failed to marshal to json: %v. Output: %v", err, string(out))
-			}
+		verInfo := resp.ClientVersion
 
-			verInfo := resp.ClientVersion
+		validateField(t, "Version", verInfo.Version)
+		validateField(t, "GitRevision", verInfo.GitRevision)
+		validateField(t, "GolangVersion", verInfo.GolangVersion)
+		validateField(t, "BuildStatus", verInfo.BuildStatus)
+		validateField(t, "GitTag", verInfo.GitTag)
+	})
+}
 
-			validateField(t, "Version", verInfo.Version)
-			validateField(t, "GitRevision", verInfo.GitRevision)
-			validateField(t, "GolangVersion", verInfo.GolangVersion)
-			validateField(t, "BuildStatus", verInfo.BuildStatus)
-			validateField(t, "GitTag", verInfo.GitTag)
-		})
+// Test that flags do not get polluted with unexpected flags
+func TestFlags(t *testing.T) {
+	runBinariesTest(t, func(t *testing.T, name string) {
+		cmd := path.Join(*releasedir, name)
+		out, err := exec.Command(cmd, "--help").Output()
+		if err != nil {
+			t.Fatalf("--help failed with error: %v. Output: %v", err, string(out))
+		}
+
+		for _, denylist := range denylistedFlags {
+			if strings.Contains(string(out), denylist) {
+				t.Fatalf("binary contains unexpected flags: %v", string(out))
+			}
+		}
+	})
+}
+
+// Test that binary sizes do not bloat
+func TestBinarySizes(t *testing.T) {
+	cases := map[string]struct {
+		minMb int64
+		maxMb int64
+	}{
+		// TODO: shrink the ranges here once the active work to reduce binary size is complete
+		// For now, having two small a range will result in lots of "merge conflicts"
+		"istioctl":        {60, 100},
+		"pilot-agent":     {30, 80},
+		"pilot-discovery": {60, 80},
+		"bug-report":      {60, 100},
 	}
+
+	runBinariesTest(t, func(t *testing.T, name string) {
+		tt, f := cases[name]
+		if !f {
+			t.Fatalf("min/max binary size not specified for %v", name)
+		}
+		cmd := path.Join(*releasedir, name)
+		fi, err := os.Stat(cmd)
+		if err != nil {
+			t.Fatal(err)
+		}
+		got := fi.Size() / (1000 * 1000)
```

Comment:
Should this be 1024 * 1024?

Reply:
Honestly not certain, I get this confused often. But per https://physics.nist.gov/cuu/Units/binary.html I think this 1024^2 would be mebibytes. using this seems to align with `ls -lh` output

Reply:
not a big deal, works for me.